src/dlsnode/util/aio/ManualContextAwaitingJob.d

```diff
@@ -0,0 +1,91 @@
+/*******************************************************************************
+
+    ManualContextAwaitingJob implementation, allowing for customisation of
+    suspend/resume operations via delegates, to the point when auto-suspension
+    of the context would be prohibited, forcing the context to suspend itself
+    at the convenient location. One usage of this behaviour would be to allow
+    resuming the context from the several external systems (e.g. from the disk
+    IO and from the network), allowing the context to suspend itself at the
+    convenient point where it's possible to distinguish if the context is resumed
+    from the network or from the AsyncIO.
+
+    copyright:
+        Copyright (c) 2016-2017 sociomantic labs GmbH. All rights reserved
+
+    License:
+        Boost Software License Version 1.0. See LICENSE.txt for details.
+
+*******************************************************************************/
+
+module dlsnode.util.aio.ManualContextAwaitingJob;
+
+import dlsnode.util.aio.ContextAwaitingJob;
+import ocean.core.Verify;
+
+/// ditto
+class ManualContextAwaitingJob: ContextAwaitingJob
```

Comment:
I think the name of this class has been confusing me, every time I've looked at this module.

The other existing `ContextAwaitingJob`-derived class is called `EventFdContextAwaitingJob`, so maybe `DelegateContextAwaitingJob` would be more consistent. I don't think those names _really_ convey what these classes are about, though.

Maybe the difficulty is that the function of these objects looks different depending whether you're looking from the perspective of the context waiting for the job vs the AIO system performing the jobs. From the POV of the AIO system `ContextAwaitingJob` is a good name (and you don't care what mechanism is used to wait/resume). From the POV of context that's waiting for the job to finish, something like `JobNotification` would make more sense. Then we'd have `EventFDJobNotification` and `DelegateJobNotification`.

You'll know better than me how much sense all of that makes ;)

Reply:
I think both of these are good points. IMHO, it makes more sense to look into this from the POV of context that's waiting for the job to finish.

Reply:
We also discussed the idea of reducing this to:
* A base class that works with two delegates.
* A derived class that provides delegates to await/trigger an event fd.